### PR TITLE
Deploy specific commits

### PR DESCRIPTION
**Pull request by esoergel**
I wanna do a little more testing before merging this, but figured I'd open it for feedback now.

Basically the idea is to deploy a specific tagged commit rather than a branch.  This was triggered by [this ticket](http://manage.dimagi.com/default.asp?216565) which points out that what gets deployed isn't always the same thing that gets tagged.  Since each server runs `git checkout origin/master`, they're not even guaranteed to end up with the same commit (someone could merge a PR at the exact wrong time).  I figured it made sense to tag at the beginning, then tell each machine to checkout that tag.
@gcapalbo @benrudolph

### CODE REVIEW

**Comment by millerdev** [comment]
:heart:

File: fab/fabfile.py [hunk 1]
```diff
@@ -165,7 +201,8 @@ def _setup_path():
     env.log_dir = posixpath.join(env.home, 'www', env.environment, 'log')
     env.releases = posixpath.join(env.root, 'releases')
     env.code_current = posixpath.join(env.root, 'current')
-    env.code_root = posixpath.join(env.releases, time.strftime('%Y-%m-%d_%H.%M', time.gmtime(time.time())))
+    timestamp = time.strftime('%Y-%m-%d_%H.%M', time.gmtime(time.time()))
```
Comment (millerdev) [inline fab/fabfile.py]: Does it matter if this time stamp is different from the one used in `CodeToDeploy`? Would it make sense to use `code_to_deploy.deploy_time` instead?

File: fab/fabfile.py [hunk 2]
```diff
@@ -611,7 +650,7 @@ def preindex_complete():
                 _execute_with_timing(stop_celery_tasks)
             _execute_with_timing(_migrate)
         else:
-            print(blue("No migration required, skipping."))
+            print blue("No migration required, skipping.")
```
Comment (millerdev) [inline fab/fabfile.py]: Why was this changed? (looks like reverting from Python 3 compatible syntax to Python 2 only.

File: fab/fabfile.py (hunk 1 above)
Comment (esoergel) [inline fab/fabfile.py]: That would make sense, yeah.  The only reason I changed this line was because it was too long.  I do agree that it'd be nice though for the deploy tag to match the code dir.  Will update.

File: fab/fabfile.py (hunk 2 above)
Comment (esoergel) [inline fab/fabfile.py]: most of the file uses print statments and there's no `from __future__ import print_function` at the top.  I just switched this for consistency, because what was happening was "print the following single-element tuple, but without a space, so it looks like a function call".

Comment (millerdev) [inline fab/fabfile.py]: :+1: to consistency

**Comment by czue** [comment]
nice this makes a lot of sense to me.

File: fab/fabfile.py [hunk 3]
```diff
@@ -463,9 +502,9 @@ def update_code(use_current_release=False):
 
     with cd(env.code_root if not use_current_release else env.code_current):
         sudo('git remote prune origin')
-        sudo('git fetch origin {}'.format(env.code_branch))
-        sudo('git checkout %(code_branch)s' % env)
-        sudo('git reset --hard origin/%(code_branch)s' % env)
+        sudo('git fetch origin --tags')
+        sudo('git checkout {}'.format(git_tag))
```
Comment (gcapalbo) [inline fab/fabfile.py]: can you checkout tags directly? i always thought you had to checkout a new branch which starts from the tag

**Comment by gcapalbo** [comment]
:+1: to the approach

File: fab/fabfile.py (hunk 2 above)
Comment (benrudolph) [inline fab/fabfile.py]: :flushed:

**Comment by benrudolph** [comment]
cool looks good to me! :rocket: ping back when you've tested it out

File: fab/fabfile.py (hunk 3 above)
Comment (esoergel) [inline fab/fabfile.py]: Yup, you'll just be in a "detached HEAD state", where you can't really save any new commits you make until you check out a new branch.  You can actually check out several types of refs in  git - branches, remote branches (`$ git checkout origin/master`), tags, or commit hashes.

Related fun fact: you can do `set --code_branch=<ref thing>` with fabric for any of those things and I think it would just work - so we could have deployed a specific commit with `$ fab production awesome_deploy --set code_branch=7d6f1ed`.

**Comment by esoergel** [comment]
Made a couple small changes and did a full staging deploy without issue.  g2m on my end.

**Comment by esoergel** [comment]
@dimagi/team-commcare-hq 
This PR makes it so we deploy a specific commit instead of a branch.  In particular, it takes the commit referenced by YOUR LOCAL COPY of the branch you're deploying (the same one you'd see if you did `$ git show master`).  This means that if you update submodules and make sure everything's hunky-dory, then some jerk comes along and merges a bad commit in, you won't deploy that commit unless you update your local master (or whatever branch you're deploying) again.  This also means that no new commits can sneak in mid-deploy (which is currently possible).

Anyways, it should just work without any noticeable differences, but in case you DO run into issues, you can checkout the old version of `fab/fabfile.py`, deploy normally, then yell at me to fix it:
```sh
$ git checkout 7d6f1ed fab/fabfile.py
$ fab environment awesome_deploy
```

File: fab/fabfile.py (hunk 3 above)
Comment (gcapalbo) [inline fab/fabfile.py]: cool, yeah knew about the rest of the stuff but i guess i've always just checked out tags with a new branch

**Comment by czue** [comment]
@esoergel just to make sure I understand this, can you clarify that the following scenarios are accurate?

1. I run deploy on master, but without having pulled the latest master first.

In this case whatever local copy of master I have (the last time I pulled) will be deployed.

2. I run a normal master deploy on my feature branch, forgetting to checkout master first.

In this case whatever local copy of master I have (the last time I pulled) will be deployed. My feature branch will not.

3. I have been working locally on a feature branch. I haven't pushed my changes up to github. I try to deploy that feature branch.

(does this fail?)

--

If these is accurate, then I'm wondering if this PR should have included checks to make sure you have the latest remote version, similar to what `rebuildstaging` does? I really like the idea of making the commit you want to deploy explicit, but It seems a little scary if there aren't any safeguards to ensure that that is in fact a good commit to be deploying (I think the safeguard would be a loud warning if you're deploying something that isn't the latest commit on the remote branch)

**Comment by esoergel** [comment]
@czue The first two scenarios are correct, but I believe scenario 3 won't fail.  That is:
```shell
$ fab production awesome_deploy --set code_branch=my-feature-branch
```
Produces the following series of git commands:
```shell
$ git rev-parse my-feature-branch
12345  # this is the commit hash of the local copy of my-feature-branch
$ git tag 2016-02-11-production-deploy -m "production deploy at 2016-02-11" 12345
$ git push origin 2016-02-11-production-deploy
```
This makes the tagged commit (12345) available on `origin` under that tag, so the deploy should work fine.

I do agree that it scenarios 1 and 2 are potentially misleading.  It should be pretty simple to add a check that that commit is the latest commit on the `origin` copy of that branch (`origin` is fetched, so it'll be up-to-date).  This could also just tag `origin/<branch>` instead of the local `<branch>`, but I think I prefer this way.  Lemme know if you agree and I'll add in a check like
> It looks like origin/\<branch> has changed since you last pulled, are you sure you want to deploy your local copy of \<branch>?

Does that sound like what you had in mind?

**Comment by czue** [comment]
Yeah that sounds perfect! :+1:
